### PR TITLE
Add a `USELESS_LET_IF_SEQ` lint

### LINKED ISSUE #15: "Unnecessary mut" cases not handled by the built-in lint
Things like

```rust
let mut x;

if p {
    x = 0u;
} else {
    x = 1u;
}

// no more mutation of x
```

which should be rewritten as

```rust
let x = if p { 0u } else { 1u };
```

### LINKED ISSUE #636: lint: `let mut foo = ..; if(..) { foo = }`
Lint things like
```rust
let mut foo = some_default;
if bar {
    foo = some_value;
}
// foo only used in non-mut context
```
which can be replaced as:
```rust
let foo = if bar {
    some_value
}
else {
    some_default
};
```
And for a use case: *[src/consts.rs#L456-L459](https://github.com/Manishearth/rust-clippy/blob/af07ccc16cbf9a736a5aba2759f296242636fecc/src/consts.rs#L456-L459)*.

### PR DESCRIPTION
**Pull request by mcarton**
Fix #15. Fix #636.

### CODE REVIEW

File: src/let_if_seq.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                    cond=snippet(cx, cond.span, "_"),
+                    then=if then.stmts.len() > 1 { " ..;" } else { "" },
+                    else=if else_.is_some() { " ..;" } else { "" },
+                    value=snippet(cx, value.span, "<value>"),
+                    default=snippet(cx, default.span, "<default>"),
+                );
+                span_lint_and_then(cx,
+                                   USELESS_LET_IF_SEQ,
+                                   span,
+                                   "`let foo;`/`if .. { foo =  ..; }` sequence detected",
```
Comment (mcarton) [inline src/let_if_seq.rs]: I’m open to suggestions for a good error message and short description (`TODO` is not descriptive enough :stuck_out_tongue_closed_eyes:).

**Comment by llogiq** [comment]
I'm curious how this handles more complex cases (E.g. with early `return`, loops or other statements inbetween definition and initialization, etc.

The reason for this being unidiomatic is that it treats `if` as a statement (which it is in C/Java/Python/..), whereas it is an expression in Rust.

**Comment by llogiq** [comment]
Regarding the error message, how about: `if _ { .. } else { .. }` is an expression. Rustic code uses it as such: [suggestion here]

**Comment by llogiq** [comment]
@mcarton do you need help on this?

**Comment by mcarton** [comment]
I completely forgot about this :sweat_smile:

**Comment by llogiq** [comment]
I periodically try to go through our PR queue and look if I can merge something or ping someone. :smile:

**Comment by llogiq** [comment]
And again, it's been more than a month. I'd still like to see more tests for this, and I'm not sure if it needs a rustup or just rebasing.

**Comment by mcarton** [comment]
@llogiq rebased and added some tests.

File: clippy_lints/src/let_if_seq.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
+use rustc::lint::*;
+use rustc::hir;
+use syntax::codemap;
+use utils::{snippet, span_lint_and_then};
+
+/// **What it does:** This lint checks for variable declarations immediatly followed by a
+/// conditionnal affectation.
```
Comment (Manishearth) [inline clippy_lints/src/let_if_seq.rs]: "conditional", otherwise r+

will merge once I get home

**Pull request merged**